src/main/scala/com/twitter/diffy/proxy/RefineHttpHeadersByLabel.scala

```diff
@@ -0,0 +1,54 @@
+package com.twitter.diffy.proxy
+
+import com.twitter.finagle.Filter
+import com.twitter.finagle.http.Request
+import com.twitter.util.Future
+import org.jboss.netty.handler.codec.http.{HttpResponse, HttpHeaders, HttpRequest}
+import scala.collection.JavaConverters._
+
+object RefineHttpHeadersByLabel {
+  val knownLabels: Seq[String] = Seq("primary", "secondary", "candidate")
+
+  def substitute(inclusions: Seq[String], exclusion: Seq[String])(req: HttpRequest) : HttpRequest = {
+
+    def orFn[T](f: T => Boolean, g: T => Boolean): T => Boolean = { t =>  f(t) || g(t) }
+    def constFalse[T](t: T): Boolean = false
+
+    def renameHeaders(headers: HttpHeaders)(nameOld: String, nameNew: String): Unit = {
+      val temp = headers.get(nameOld)
+      headers.remove(nameOld)
+      headers.set(nameNew, temp)
+    }
+
+
+    val removeExclusion  = exclusion.
+      map(_+"_").
+      map( x => (_:String).startsWith(x)).
+      foldLeft(constFalse[String]_)(orFn)
+
+    val reqOut = Request.apply(req)
+    val headers: HttpHeaders = reqOut.headers
+    //k => Option[Function[k => k]]
+    val incomingHeaders = headers.names().asScala.toSet
+
+    val rewrites = for {
+      name <- incomingHeaders
+      inclusion <- inclusions if name.startsWith(inclusion + "_")
+    } yield name->name.substring(inclusion.length+1)
+
+    incomingHeaders.filter( removeExclusion).foreach( headers.remove )
+    rewrites.foreach((renameHeaders(headers) _ ).tupled)
+
+    reqOut
+  }
+
+  def apply(label:String): Filter[HttpRequest, HttpResponse, HttpRequest, HttpResponse] = {
+    def mkFilter(inclusions: Seq[String], exclusion: Seq[String])
+                (req:HttpRequest, filter: (HttpRequest => Future[HttpResponse]) ) =
+      filter(substitute(inclusions, exclusion)(req))
+
+    val (inclusions, exclusion) = knownLabels.partition(_.equalsIgnoreCase(label))
```

Comment:
The expected behavior is that "inclusions" should always be equal to "Seq(label)". You can simplify this code by using that observation and defining "exclusions" as "knownLabels filter { !_.equalsIgnoreCase(label) }" and eliminating inclusions by directly using label instead.